sympy/solvers/ode.py

```diff
@@ -1482,6 +1491,245 @@ def _test_term(coeff, order):
+        sol = Eq(func, sol)
+
+    # if sol is None then we can try for series solution
+    if sol is None:
+        raise NotImplementedError("The given ODE " + str(eq) + " cannot be solved by"
+            + " the hypergeometric method")
+
+    return sol
+
+def match_2nd_2F1_hypergeometric(I, k, sing_point, func):
+
+    from sympy import factor
+    x = func.args[0]
+    a = Wild("a")
+    b = Wild("b")
+    c = Wild("c")
+    t = Wild("t")
+    s = Wild("s")
+    alpha = Wild("alpha")
+    beta = Wild("beta")
+    gamma = Wild("gamma")
+    delta = Wild("delta")
+
+    rn = {'type':None}
+    # I0 of the standerd 2F1 equation.
+    I0 = ((a-b+1)*(a-b-1)*x**2 + 2*((1-a-b)*c + 2*a*b)*x + c*(c-2))/(4*x**2*(x-1)**2)
+    if sing_point != [0, 1]:
+        # If singular point is [0, 1] then we have standerd equation.
+        eqs = []
+        sing_eqs = [-beta/alpha, -delta/gamma, (delta-beta)/(alpha-gamma)]
+        # making equations for the finding the mobius transformation
+        for i in range(3):
+            if i<len(sing_point):
+                eqs.append(Eq(sing_eqs[i], sing_point[i]))
+            else:
+                eqs.append(Eq(1/sing_eqs[i], 0))
+        # solving above equations for the mobius transformation
+        _beta = -alpha*sing_point[0]
+        _delta = -gamma*sing_point[1]
+        _gamma = alpha
+        if len(sing_point) == 3:
+            _gamma = (_beta + sing_point[2]*alpha)/(sing_point[2] - sing_point[1])
+        mob = (alpha*x + beta)/(gamma*x + delta)
+        mob = mob.subs(beta, _beta)
+        mob = mob.subs(delta, _delta)
+        mob = mob.subs(gamma, _gamma)
+        mob = cancel(mob)
+        t = (beta - delta*x)/(gamma*x - alpha)
+        t = cancel(((t.subs(beta, _beta)).subs(delta, _delta)).subs(gamma, _gamma))
+    else:
+        mob = x
+        t = x
+
+    # applying mobius transformation in I to make it into I0.
+    I = I.subs(x, t)
+    I = I*(t.diff(x))**2
+    I = factor(I)
+    dict_I = {x**2:0, x:0, 1:0}
+    I0_num, I0_dem = I0.as_numer_denom()
+    # collecting coeff of (x**2, x), of the standerd equation.
+    dict_I0 = {x**2:(a-b)**2 - 1, x:2*((1-a-b)*c + 2*a*b), 1:c*(c-2)}
+    # collecting coeff of (x**2, x) from I0 of the given equation.
+    dict_I.update(collect(expand(cancel(I*I0_dem)), [x**2, x], evaluate=False))
+    eqs = []
+    # We are comparing the coeff of powers of different x, for finding the values of
+    # parameters of standerd equation.
+    for key in [x**2, x, 1]:
+        eqs.append(Eq(dict_I[key], dict_I0[key]))
+
+    # We can have many possible roots for the equation.
+    # I am selecting the root on the basis that when we have
+    # standard equation eq = x*(x-1)*f(x).diff(x, 2) + ((a+b+1)*x-c)*f(x).diff(x) + a*b*f(x)
+    # then root should be a, b, c.
+    _c = list(ordered(roots(eqs[2], c)))[0]
+    if not _c.has(Symbol):
+        _c = min(list(roots(eqs[2], c)))
+    s = list(ordered(roots(eqs[0].subs(a-b, s), s)))[-1]
```

Comment:
I don't understand this comment. Can you explain this in detail?

Reply:
Above you have:
```python
    dict_I0 = {x**2:(a-b)**2 - 1, x:2*((1-a-b)*c + 2*a*b), 1:c*(c-2)}
```
Change that to something like (untested)
```python
    dict_I0 = {x**2: d**2 - 1, x:2*((1-e)*c + 2*(d+e)*(d-e)), 1:c*(c-2)}
```
and then solve for d in the first then e in the second then sub d, e for a and b in the solution.

Really though if you know which root you are taking from these equations you can hard-code the solution. No need to actually call roots at runtime: run roots once now to get the formula and then use that formula in the code.

Reply:
I have done this but due to this, some problem is coming.
```
In [9]: a = Symbol('a', positive=True)

In [10]: b = Symbol('b', positive=True)

In [12]: sqrt(a**2 + b**2 + 2*a*b)
Out[12]: 
   _________________
  ╱  2            2 
╲╱  a  + 2⋅a⋅b + b  

In [13]: a>0
Out[13]: True

In [14]: b>0
Out[14]: True


```
The expression is not converting to (a+b) and it will cause a problem in the solution and for the equation having symbols very complex solution is coming due to this.

Reply:
I don't quite understand what you mean. You can simplify that example with `factor`. However the point is that you can just do this step by step: first solve for c using quadratic formula (f is rhs):
```julia
In [20]: roots(c*(c-2)-f, c)                                                                             
Out[20]: 
⎧      _______       _______       ⎫
⎨1 - ╲╱ f + 1 : 1, ╲╱ f + 1  + 1: 1⎬
⎩                                  ⎭
```
Choose which of those you want and put that explicitly in the code (no need to call roots at runtime).

Then you can solve for a+b = d from:
```julia
In [22]: roots(d**2 - 1 - e, d)                                                                          
Out[22]: 
⎧   _______       _______   ⎫
⎨-╲╱ e + 1 : 1, ╲╱ e + 1 : 1⎬
⎩                           ⎭
```
Again choose one and put the formula in the code. That gives us `a+b` so then we get `b` from
```julia
In [45]: eq = 2*((1-a-b)*c + 2*a*b) - g                                                                  

In [46]: eq.subs(a, d-b).expand().collect(b)                                                             
Out[46]: 
     2                          
- 4⋅b  + 4⋅b⋅d - 2⋅c⋅d + 2⋅c - g

In [47]: roots(_, b)                                                                                     
Out[47]: 
⎧       _______________________            _______________________   ⎫
⎪      ╱                 2                ╱                 2        ⎪
⎨d   ╲╱  -2⋅c⋅d + 2⋅c + d  - g      d   ╲╱  -2⋅c⋅d + 2⋅c + d  - g    ⎬
⎪─ - ──────────────────────────: 1, ─ + ──────────────────────────: 1⎪
⎩2               2                  2               2                ⎭
```
Again choose one of those.

Once you've made those three choices the whole thing can be hard coded so no need to call roots.